src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,497 @@
+                image,
+                version,
+                self.registry_name,
+            )
+
+            try:
+                manifest_inspect_cmd = [
+                    str(shutil.which("docker")),
+                    "manifest",
+                    "inspect",
+                    image_path,
+                ]
+
+                output = call_subprocess_raise_output(manifest_inspect_cmd)
+
+                if output is not None:
+                    return self, namespace
+
+            except CLIError as error:
+                if "manifest unknown" in str(error):
+                    continue
+                logger.warning(
+                    (
+                        "Failed to contact source registry %s. "
+                        "Make sure you run docker login on this registry "
+                        "before running the aosm command."
+                    ),
+                    self.registry_name,
+                )
+                logger.debug(error, exc_info=True)
+                raise ClientRequestError(
+                    f"Failed to contact source registry {self.registry_name} "
+                    "Make sure you run docker login on this registry "
+                    "before running the aosm command."
+                ) from error
+        return None, None
+
+
+class AzureContainerRegistry(ContainerRegistry):
+    """
+    A class to represent an Azure Container Registry and handle all ACR operations.
+    """
+
+    # TODO: Have I tested this code path?
+    def _login(self) -> None:
+        """
+        Log in to the source registry using the Azure CLI.
+        Uses the CLI user's context to log in to the source registry.
+        """
+
+        message = f"Logging into source registry {self.registry_name}"
+        logger.info(message)
+
+        try:
+            acr_source_login_cmd = [
+                str(shutil.which("az")),
+                "acr",
+                "login",
+                "--name",
+                self.registry_name,
+            ]
+            call_subprocess_raise_output(acr_source_login_cmd)
+
+        except CLIError as error:
+            logger.debug(error, exc_info=True)
+            raise ClientRequestError(
+                f"Failed to log into registry {self.registry_name}"
+            ) from error
+
+    def pull_image_to_local_registry(self, source_image: str) -> None:
+        """
+        Pull image to local registry using docker pull. Requires docker.
+
+        :param: source_image: source docker image name e.g.
+            uploadacr.azurecr.io/samples/nginx:stable
+        """
+
+        self._login()
+        try:
+            super().pull_image_to_local_registry(source_image)
+        except BadRequestError as error:
+            raise error
+        finally:
+            logger.info("Logging out of source registry %s", self.registry_name)
+
+            docker_logout_cmd = [
+                str(shutil.which("docker")),
```

Comment:
Why are we using `docker` to logout? If this is correct, I'd add a comment to explain why.